### PR TITLE
Finish Week 4 Tasks

### PR DESCRIPTION
**Pull request by shiyi001**
finish unit test tutorial and Calendar task(pass all unit tests).

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,73 @@
...
+        for (String attendee : attendees) {
+            for (Event e : events) {
+                if (hasAdded.contains(e)) continue;
+                if (e.getAttendees().contains(attendee)) {
+                    TimeRange when = e.getWhen();
+                    attendeeMeetingTimeRanges.add(when);
+                    hasAdded.add(when);
+                }
+            }
+        }
```
Comment (Time1ess) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:83]: Maybe you can switch the outer forloop and inner forloop.

```
for (Event) {
  for(attendee) {
    ...
    if (contains) {
      add();
      break;
  }
}
```

This way you can save one `hashset`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,73 @@
...
+        int idx = 0;
+        while (idx < attendeeMeetingTimeRanges.size()) {
+            while ((idx < attendeeMeetingTimeRanges.size()) && 
+                (start >= attendeeMeetingTimeRanges.get(idx).start())) {
+                start = Math.max(start, attendeeMeetingTimeRanges.get(idx).end());
+                idx++;
+            }
+            if (idx >= attendeeMeetingTimeRanges.size()) break;
+
+            if (!attendeeMeetingTimeRanges.get(idx).contains(start)) {
```
Comment (Time1ess) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: CMIIW, the conditions to reach this if-statement is

1. idx < size()
2. start < attendeeMeetingTimeRanges.get(idx).start()

So why do you need this if-statement?

**Review by Time1ess — CHANGES_REQUESTED** [review]

File: walkthroughs/week-4-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 3]
```diff
@@ -22,12 +22,22 @@
...
+    }
+
+    @Test
+    public void testGreetingTrimsWhitespace() {
+        Greeter greeter = new Greeter();
+
+        String greeting = greeter.greet("   Ada   ");
+
+        // Whitespace should be trimmed
+        Assert.assertEquals("Hello Ada", greeting);
```
[the remark this answers is not in the document]
Comment (shiyi001) [reply walkthroughs/week-4-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:41]: I will modify it.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,73 @@
...
+        }
+        if ((start < TimeRange.END_OF_DAY) && 
+            (TimeRange.END_OF_DAY - start >= request.getDuration())) {
+            availableTimeRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+        }
+
+        return availableTimeRanges;
+    }
+
+    private ArrayList<TimeRange> getAttendeeMeetingTimeRanges(
```
[the remark this answers is not in the document]
Comment (shiyi001) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:67]: I will check it and modify it if so.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,73 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
[the remark this answers is not in the document]
Comment (shiyi001) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: This interface is pre-defined, I think it is best not to modify.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (shiyi001) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:83]: Great suggestion!

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (shiyi001) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: There is no need for if-statement here.

**Comment by shiyi001** [comment]
Modify according to comment

**Comment by shiyi001** [comment]
Modify according to comment

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,81 @@
...
+import java.util.HashSet;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    
+    /**
+     * Return available time period for a meeting request
```
Comment (Time1ess) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: Nit: Use third-person verbs. e.g. Returns xxx.

**Review by Time1ess — APPROVED** [review]

**Pull request merged**